### PR TITLE
Add PATCH Contact Details Endpoint

### PR DESCRIPTION
**Pull request by LBHCallumM**
## Ticket

[https://hackney.atlassian.net/browse/MR-761?atlOrigin=eyJpIjoiMGU1N2E5NTAzODU3NGViNTgzNTBmZTg3YzQ1YjVhM2QiLCJwIjoiaiJ9](https://hackney.atlassian.net/browse/MR-761?atlOrigin=eyJpIjoiMGU1N2E5NTAzODU3NGViNTgzNTBmZTg3YzQ1YjVhM2QiLCJwIjoiaiJ9)

## Summary of Changes

Add a PATCH endpoint to allow contact details to be updated. Contact details are stored in DynamoDb, with a hash key (personId) and a sort key (contactDetailId). I wasn't sure how best to format the URL. I am open to new suggestions.

Route: [PATCH] ` /contactDetails/{personId}/update/{contactDetailId}`

The request object currently only accepts `ContactInformation`, as this is what needs to be updated.

Request object: 
```json
{
  "ContactInformation": {}
}
```

I have not implemented optimistic locking yet, which is used by most other patch PATCH endpoints, because this added too much complexity to this PR. Especially considering this PR has V1 and V2 routes, and multiple instances of the `ContactDetailsEntity` object for each version. I will look at adding this in a later PR.

### Other changes

- I have followed the existing pattern used in the AssetAPI, by publishing an SNS event.

### CODE REVIEW

**Comment by Asura5** [comment]
In a restful API, usually the route signifies the resources, and the Http verb defines the a action. We know it's an update, because it's a patch. Having update in the route is A bit RPc

File: ContactDetailsApi/V2/Controllers/ContactDetailsController.cs [hunk 1]
```diff
@@ -84,5 +88,19 @@ public async Task<IActionResult> CreateContact([FromBody] ContactDetailsRequestO
                 return BadRequest(e.ConstructResponse());
             }
         }
+
+        [HttpPatch]
+        [Route("{personId}/update/{contactDetailId}")]
```
Comment (humulla) [inline ContactDetailsApi/V2/Controllers/ContactDetailsController.cs:93]: Maybe we should change the route to `/{contactDetailId}/person/{personId}`

Comment (LBHCallumM) [reply ContactDetailsApi/V2/Controllers/ContactDetailsController.cs:93]: Yeah, I prefer this

File: ContactDetailsApi/V2/Gateways/ContactDetailsDynamoDbGateway.cs [hunk 2]
```diff
@@ -49,16 +54,41 @@ public async Task<List<ContactDetails>> GetContactDetailsByTargetId(ContactQuery
...
+
+            var updaterResponse = _updater.UpdateEntity(
+                existingContactDetails,
+                requestBody,
+                request.ToDatabase()
+            );
+
+            if (updaterResponse.NewValues.Any())
+            {
+                _logger.LogDebug("Calling IDynamoDBContext.SaveAsync to update {ContactId} {PersonId}", query.ContactDetailId, query.PersonId);
```
Comment (humulla) [inline ContactDetailsApi/V2/Gateways/ContactDetailsDynamoDbGateway.cs:73]:
```suggestion
                _logger.LogDebug("Calling IDynamoDBContext.SaveAsync to update  with contactId:{ContactId} and personId:{PersonId}", query.ContactDetailId, query.PersonId);
```

**Review by humulla — APPROVED** [review]
Small suggestion on logging but the rest looks good

File: ContactDetailsApi/V2/Gateways/ContactDetailsDynamoDbGateway.cs (hunk 2 above)
Comment (LBHCallumM) [reply ContactDetailsApi/V2/Gateways/ContactDetailsDynamoDbGateway.cs:73]: Not sure this is necessary with a templated log string, but I can add it

**Pull request merged**